File: C++_Base_PT4/OOP3Behav2.cpp
#include "pt4.h"
#include <algorithm>
#include <list>

using namespace std;

class Observer;

class Subject
{
public:
    virtual void Attach(Observer* observ) = 0;
    virtual void Detach(Observer* observ) = 0;
    virtual void SetInfo(string info) = 0;
};

class Observer
{
public:
    virtual void OnInfo(Subject* sender, string info) = 0;
};

class ConcreteSubject: public Subject
{
public:
	void Attach(Observer* pObserver) 
	{
		m_ListObserver.push_back(pObserver);
	}         

	void Detach(Observer* pObserver) 
	{
		list<Observer*>::iterator iter;
		iter = find(m_ListObserver.begin(), m_ListObserver.end(), pObserver);
		if (iter != m_ListObserver.end()) 
		{
			m_ListObserver.erase(iter);
		}
	}         

	void SetInfo(string info) 
	{
		list<Observer*>::iterator iter = m_ListObserver.begin();
		for (; iter != m_ListObserver.end();)
		{
			list<Observer*>::iterator l = iter;
			l++;
			(*iter)->OnInfo(this, info);
			iter = l;
		}	
	}    

private:
	list<Observer*>    m_ListObserver;
};

class ConcreteObserver : public Observer
{
public:	
	ConcreteObserver(char detachInfo){
		this->detachInfo = detachInfo;
		this->log = "";
	}

	void OnInfo(Subject* sender, string info)
	{
		this->log = this->log + info;

		int n = log.length() - 1;
		if(log[n] == detachInfo)
		{
			sender->Detach(this);
		}
	}
	
	void Attach(Subject* pSubject) 
	{
		pSubject->Attach(this);
	}         

	void Detach(Subject* pSubject) 
	{
		pSubject->Detach(this);
	}   
	
	string GetLog()
	{
		return log;	
	}	
	
private:
	string log;
	char detachInfo;
};

void Solve()
{
    Task("OOP3Behav2");
	int N;
	int K;
	pt >> N >> K;
	string str;

	ConcreteSubject* subj1 = new ConcreteSubject;
	ConcreteSubject* subj2 = new ConcreteSubject;	
	
	ConcreteObserver* obser[10];
	for (int i = 0; i < N; i++) {
		obser[i] = new ConcreteObserver('a' + i);
		obser[i]->Attach(subj1);
		obser[i]->Attach(subj2);
	}

	for (int i = 0; i < K; i++)
	{
		pt >> str;
		if(str[0]=='1')
			subj1->SetInfo(str);
		if(str[0]=='2') 
			subj2->SetInfo(str);
	}

	for (int i = 0; i < N; i++)
	{
		ShowLine(obser[i]->GetLog()) ;
		pt << obser[i]->GetLog();
	}
}

File: C#_Base_PT4/OOP3Behav2.cs
// File: "OOP3Behav2"
using PT4;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PT4Tasks
{
    public class MyTask : PT
    {
        public abstract class Subject
        {
            public abstract void Attach(Observer observ);
            public abstract void Detach(Observer observ);
            public abstract void SetInfo(string info);
        }

        // Implement the ConcreteSubject descendant class

        public class ConcreteSubject : Subject
        {
            List<Observer> m_ListObserver = new List<Observer>();

            public override void Attach(Observer pObserver)
            {
                m_ListObserver.Add(pObserver);
            }

            public override void Detach(Observer observ)
            {
                m_ListObserver.Remove(observ);
            }

            public override void SetInfo(string info)
            {
                for(int i = m_ListObserver.Count - 1; i >= 0; i--)
                {
                    m_ListObserver[i].OnInfo(this, info);
                }
            }
        }
        
        public abstract class Observer
        {
            public abstract void OnInfo(Subject sender, string info);
        }

        // Implement the ConcreteObserver descendant class
        public class ConcreteObserver : Observer
        {
            string log;
            char detachInfo;
            public ConcreteObserver(char datachInfo){
                this.detachInfo = datachInfo;
                this.log = null;
            }

            public override void OnInfo(Subject sender, string info)
            {
                this.log = this.log + info;

                int n = log.Length - 1;
                if(log[n] == detachInfo)
                {
                    sender.Detach(this);
                }
            }

            public void Attach(Subject pSubject)
            {
                pSubject.Attach(this);
            }
            public void Detach(Subject pSubject)
            {
                pSubject.Detach(this);
            }
            public string GetLog(){
                return log;
            }
        }

        public static void Solve()
        {
            Task("OOP3Behav2");
            int N = GetInt();
            int K = GetInt();
            string str;
            
            ConcreteSubject subj1 = new ConcreteSubject();
            ConcreteSubject subj2 = new ConcreteSubject();

            ConcreteObserver[] obser = new ConcreteObserver[10];
            for(int i = 0; i < N; i++)
            {
                obser[i] = new ConcreteObserver(Convert.ToChar('a' + i));
                obser[i].Attach(subj1);
                obser[i].Attach(subj2);
            }

            for(int i = 0; i < K; i++)
            {
                str = GetString();
                if(str[0] == '1')
                    subj1.SetInfo(str);
                if(str[0] == '2')
                    subj2.SetInfo(str);
            }

            for(int i = 0; i < N; i++)
            {
                Put(obser[i].GetLog());
            }
        }
    }
}
